### PR TITLE
Group G,Group D

### PR DESCRIPTION
**Pull request by harishsdev**
test_five_crop
test_ten_crop
test_max_value
test_linear_transformation
test_autoaugment
test_center_crop
test_center_crop_2
test_color_jitter

### CODE REVIEW

**Comment by harishsdev** [comment]
PR for :https://github.com/pytorch/vision/issues/3945

**Comment by NicolasHug** [comment]
Thanks @harishsdev , there are failures, some of theme being undefined names: https://app.circleci.com/pipelines/github/pytorch/vision/8695/workflows/230c01ae-5eaa-4a88-9d4e-e480d9be8dfc/jobs/631032

Did you manage to run the tests locally?

**Comment by harishsdev** [comment]
@NicolasHug  Thanks for your support,I'm able run tests in my local PC,I found no error there,

Please help to resolve below error

assertEqual(output_pil.size()[1:3], crop_size,
                        "image_size: {} crop_size: {}".format(input_image_size, crop_size))
E           NameError: name 'assertEqual' is not defined

test/test_transforms.py:1948: NameError

**Comment by harishsdev** [comment]
My doubt is how to write assertEqual for formating expression!

**Comment by NicolasHug** [comment]
> I'm able run tests in my local PC,I found no error there,

There's probably something wrong then because otherwise you would have gotten the same issue as above. did you run `pytest test/test_transforms.py` ?

> My doubt is how to write assertEqual for formating expression!

you can juste write `assert a == b`, without an error message.

**Comment by harishsdev** [comment]
Hi @NicolasHug  Error is coming from file which i have not modified,please guide me how to proceed

(practice) PS D:\contrib\19283975\exp\vision> pytest test/test_transforms.py -vvv -k test_five_crop
==================================================== test session starts ====================================================
platform win32 -- Python 3.8.8, pytest-6.2.3, py-1.10.0, pluggy-0.13.1 -- c:\users\harish\anaconda3\envs\practice\python.exe
cachedir: .pytest_cache
rootdir: D:\contrib\19283975\exp\vision
plugins: cov-2.11.1, pythonpath-0.7.3, typeguard-2.12.0
collected 0 items / 1 error

========================================================== ERRORS ===========================================================
_________________________________________ ERROR collecting test/test_transforms.py __________________________________________
test\test_transforms.py:25: in <module>
    from common_utils import cycle_over, int_dtypes, float_dtypes
test\common_utils.py:20: in <module>
    from _assert_utils import assert_equal
test\_assert_utils.py:11: in <module>
    assert_equal = functools.partial(torch.testing.assert_close, rtol=0, atol=0)
E   AttributeError: module 'torch.testing' has no attribute 'assert_close'
===================================================== warnings summary ======================================================
c:\users\harish\anaconda3\envs\practice\lib\site-packages\pyreadline\py3k_compat.py:8
  c:\users\harish\anaconda3\envs\practice\lib\site-packages\pyreadline\py3k_compat.py:8: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.9 it will stop working
    return isinstance(x, collections.Callable)

torchvision\__init__.py:26
  D:\contrib\19283975\exp\vision\torchvision\__init__.py:26: UserWarning: You are importing torchvision within its own root folder (D:\contrib\19283975\exp\vision). This is not expected to work and may give errors. Please exit the torchvision project source and relaunch your python interpreter.
    warnings.warn(message.format(os.getcwd()))

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================================================== short test summary info ==================================================
ERROR test/test_transforms.py - AttributeError: module 'torch.testing' has no attribute 'assert_close'
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
=============================================== 2 warnings, 1 error in 1.31s ================================================

**Comment by harishsdev** [comment]
Tested in Local PC

pytest test/test_transforms.py -vvv -k test_five_crop-passed
pytest test/test_transforms.py -vvv -k test_ten_crop-error
pytest test/test_transforms.py -vvv -k test_linear_transformation-failed
pytest test/test_transforms.py -vvv -k test_autoaugment -passed
pytest test/test_transforms.py -vvv -k test_center_crop -failed
pytest test/test_transforms.py -vvv -k test_center_crop_2 -failed
pytest test/test_transforms.py -vvv -k test_color_jitter-passed

**Comment by harishsdev** [comment]
one function is failing as per my observation

pytest test/test_transforms.py -vvv -k test_five_crop-passed
pytest test/test_transforms.py -vvv -k test_ten_crop-passed
pytest test/test_transforms.py -vvv -k test_linear_transformation-failed
pytest test/test_transforms.py -vvv -k test_autoaugment -passed
pytest test/test_transforms.py -vvv -k test_center_crop -passed
pytest test/test_transforms.py -vvv -k test_center_crop_2 -passed
pytest test/test_transforms.py -vvv -k test_color_jitter-passed

**Comment by harishsdev** [comment]
Group G,Group D

PR for https://github.com/pytorch/vision/issues/3945

**Comment by harishsdev** [comment]
@NicolasHug please give me inputs if any further modifications required

File: test/test_transforms.py [hunk 1]
```diff
@@ -1997,5 +1757,245 @@ def test_random_order():
...
+                               msg="cov not close to 1")
+    torch.testing.assert_close(mean / num_samples, 0, rtol=1e-3, atol=1e-8, check_dtype=False,
+                               msg="mean not close to 0")
+
+    # Checking if LinearTransformation can be printed as string
+    whitening.__repr__()
+
+
+def test_max_value():
+    for dtype in int_dtypes():
```
Comment (NicolasHug) [inline test/test_transforms.py:1794]: we could parametrize over dtype here

File: test/test_transforms.py [hunk 2]
```diff
@@ -1997,5 +1757,245 @@ def test_random_order():
...
+
+    assert len(results) == 10
+    assert results == expected_output
+
+
+def test_five_crop():
+    to_pil_image = transforms.ToPILImage()
+    h = random.randint(5, 25)
+    w = random.randint(5, 25)
+    for single_dim in [True, False]:
```
Comment (NicolasHug) [inline test/test_transforms.py:1845]: we can parametrize over single_dim

File: test/test_transforms.py [hunk 3]
```diff
@@ -1997,5 +1757,245 @@ def test_random_order():
...
+
+    # Since height is independent of width, we can ignore images with odd height and even width and vice-versa.
+    input_image_sizes = [even_image_size, odd_image_size]
+
+    # Get different crop sizes
+    delta = random.choice((1, 3, 5))
+    crop_size_delta = [-2 * delta, -delta, 0, delta, 2 * delta]
+    crop_size_params = itertools.product(input_image_sizes, crop_size_delta, crop_size_delta)
+
+    for (input_image_size, delta_height, delta_width) in crop_size_params:
```
Comment (NicolasHug) [inline test/test_transforms.py:1935]: let's parametrize over these 3 variables as well.

File: test/test_transforms.py [hunk 4]
```diff
@@ -1997,5 +1757,245 @@ def test_random_order():
...
+    ])(img)
+    sum2 = result.sum()
+    assert sum2 > 0
+    assert sum2 > sum1
+
+
+def test_center_crop_2():
+    """ Tests when center crop size is larger than image size, along any dimension"""
+    even_image_size = (random.randint(10, 32) * 2, random.randint(10, 32) * 2)
+    odd_image_size = (even_image_size[0] + 1, even_image_size[1] + 1)
```
Comment (NicolasHug) [inline test/test_transforms.py:1925]: to parametrize over `input_images_sizes` and preserve the current logic, we can  do something like:

```py
@pytest.mark.parametrize('odd_size', (True, False))
def test_center_crop_2(odd_size, ...):
    input_image_size = (random.randint(10, 32) * 2, random.randint(10, 32) * 2)
    if odd_size:
        input_image_size = (input_image_size[0] + 1, input_image_size[1] + 1)
```

**Review by NicolasHug — COMMENTED** [review]
Thanks @harishsdev , we can parametrize a bit further but this looks quite good already

File: test/test_transforms.py [hunk 5]
```diff
@@ -1997,5 +1757,245 @@ def test_random_order():
...
+    torch.testing.assert_close(cov / num_samples, np.identity(1), rtol=2e-3, atol=1e-8, check_dtype=False,
+                               msg="cov not close to 1")
+    torch.testing.assert_close(mean / num_samples, 0, rtol=1e-3, atol=1e-8, check_dtype=False,
+                               msg="mean not close to 0")
+
+    # Checking if LinearTransformation can be printed as string
+    whitening.__repr__()
+
+
+@pytest.mark.parametrize('dtype', 'int_dtypes')
```
Comment (NicolasHug) [inline test/test_transforms.py:1793]: This should be
```suggestion
@pytest.mark.parametrize('dtype', int_dtypes())
```

Are you running the tests locally @harishsdev ?

Comment (harishsdev) [reply test/test_transforms.py:1793]: I'm facing same issue as previous ,assert_close is rasing error

**Comment by harishsdev** [comment]
(myenv) PS D:\contrib\19283975\exp\vision> pytest test/test_transforms.py -vvv -k test_five_crop
==================================================== test session starts ====================================================
platform win32 -- Python 3.8.8, pytest-6.2.3, py-1.10.0, pluggy-0.13.1 -- G:\installconda\python.exe
cachedir: .pytest_cache
rootdir: D:\contrib\19283975\exp\vision
plugins: anyio-2.2.0
collected 0 items / 1 error

========================================================== ERRORS ===========================================================
_________________________________________ ERROR collecting test/test_transforms.py __________________________________________
test\test_transforms.py:25: in <module>
    from common_utils import cycle_over, int_dtypes, float_dtypes
test\common_utils.py:20: in <module>
    from _assert_utils import assert_equal
test\_assert_utils.py:11: in <module>
    assert_equal = functools.partial(torch.testing.assert_close, rtol=0, atol=0)
E   AttributeError: module 'torch.testing' has no attribute 'assert_close'
===================================================== warnings summary ======================================================
G:\installconda\lib\site-packages\pyreadline\py3k_compat.py:8
  G:\installconda\lib\site-packages\pyreadline\py3k_compat.py:8: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.9 it will stop working
    return isinstance(x, collections.Callable)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================================================== short test summary info ==================================================
ERROR test/test_transforms.py - AttributeError: module 'torch.testing' has no attribute 'assert_close'
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
================================================ 1 warning, 1 error in 1.59s ================================================

**Comment by NicolasHug** [comment]
I already replied to you about this issue here: https://github.com/pytorch/vision/pull/3975#issuecomment-855368203

**Comment by harishsdev** [comment]
@NicolasHug Looks like all the tests passed finally,Can you please help me to merge this change

**Comment by NicolasHug** [comment]
Thanks @harishsdev , I took care of the  prameterization of test_center_crop_2. I'll merge when green

**Review by NicolasHug — APPROVED** [review]

**Pull request merged**